synapse/util/caches/treecache.py

```diff
@@ -57,4 +60,33 @@ def pop(self, key, default=None):
                 break
             node_and_keys[i+1][0].pop(k)
 
+        popped, cnt = _strip_and_count_entires(popped)
+        self.size -= cnt
         return popped
```

Comment:
If you are going to enumerate the popped tree here then you might as well return a list of popped entries for the LruCache to remove from its linked list.